src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt

```diff
@@ -516,44 +519,354 @@ abstract class ManagedStrategy(
         return 0
     }
 
-    /**
-     * This method is executed before a shared variable read operation.
-     * @param iThread the number of the executed thread according to the [scenario][ExecutionScenario].
-     * @param codeLocation the byte-code location identifier of this operation.
-     */
-    internal fun beforeSharedVariableRead(iThread: Int, codeLocation: Int, tracePoint: ReadTracePoint?) {
+    // == UTILITY METHODS ==
+
+    override fun lock(monitor: Any, codeLocation: Int): Unit = runInIgnoredSection {
+        val tracePoint = if (collectTrace) {
+            val iThread = currentThread
+            MonitorEnterTracePoint(iThread, currentActorId[iThread], callStackTrace[iThread], CodeLocations.stackTrace(codeLocation))
+        } else {
+            null
+        }
+        beforeLockAcquire(codeLocation, tracePoint, monitor)
+    }
+
+    override fun unlock(monitor: Any, codeLocation: Int): Unit = runInIgnoredSection {
+        val tracePoint = if (collectTrace) {
+            val iThread = currentThread
+            MonitorExitTracePoint(iThread, currentActorId[iThread], callStackTrace[iThread], CodeLocations.stackTrace(codeLocation))
+        } else {
+            null
+        }
```

Comment:
Here and throughout the `ManagedStrategy` class it is perhaps worth to consider a refactoring, where instead of doing `if (collectTrace)` every time, 
we set up a property `traceCollector: TraceCollector?`, and then just use it as:
```
traceCollector?.beforeLockRelease(monitor, codeLocation)
```

But I think this can be done in a separate PR.

Reply:
Ok, nice point, let's do it a little bit later in a separate PR